slapo/build.py

```diff
@@ -0,0 +1,247 @@
+            orig_shape = (
+                param.orig_shape if hasattr(param, "orig_shape") else param.shape
+            )
+            new_param = nn.Parameter(
+                torch.empty(orig_shape, dtype=param.dtype, device=local_rank)
+            )
+            sch.mod.register_parameter(
+                param_name,
+                new_param,
+            )
+            if hasattr(param, "replicated_param") and param.replicated_param:
+                new_param.replicated_param = True
+
+        # Use original shape to initialize parameters.
+        if global_rank == curr_stage_devices[0] and num_params > 0:
+            # only the first device in the PP group needs to initialize the weights
+            _init_module(sch)
+
+        # Broadcast complete params from rank 0 to make sure all the TP+DP ranks
+        # take the same params.
+        if dist.is_initialized():
+            curr_stage_group = stage_groups[curr_part_idx]
+            for _, param in sch.mod.named_parameters(recurse=False):
+                dist.broadcast(param, src=curr_stage_devices[0], group=curr_stage_group)
+
+        # Only keep the partition for this device for sharded params.
+        tp_rank = sch.rank
+        cnt_shard = 0
+        for param_name, param in sch.mod.named_parameters(recurse=False):
+            is_found = False
+            for idx, new_size in enumerate(new_param_shapes[param_name]):
+                if new_size != param.shape[idx]:
+                    assert not is_found, "Cannot have two sharded dimensions!"
+                    sharded_size = new_size
+                    axis = idx
+                    is_found = True
+            if is_found:
+                cnt_shard += 1
+                sharded_param = param.detach().split(sharded_size, dim=axis)[tp_rank]
+                sharded_param = sharded_param.contiguous()
+                new_param = nn.Parameter(sharded_param)
+                new_param.tensor_model_parallel = True
+                sch.mod.register_parameter(param_name, new_param)
+
+        for subsch in sch.child.values():
+            ret = _consolidate_and_broadcast(subsch)
+            num_params += ret[0]
+            cnt_shard += ret[1]
+
+        return num_params, cnt_shard
+
+    if cnt_meta != 0 or cnt_materialized != 0:
+        num_params, cnt_shard = _consolidate_and_broadcast(sch)
+
+    logger.info(
+        "Finished consolidating %d parameter tensors with %d being sharded",
+        num_params,
+        cnt_shard,
+    )
+
+    gc.collect()
+    torch.cuda.empty_cache()
+
+    return sch
+
+
+def init_target_engine(sch, target, **kwargs):
+    """Initialize the runtime engine for a specific target framework."""
+    init_engine_fn = get_dialect_cls("runtime_engine", target, allow_none=True)
+    return init_engine_fn(
+        sch,
+        **kwargs,
+    )
+
+
+def build(
+    sch: Schedule,
+    target=None,
+    init_weights: Optional[Union[bool, Callable]] = True,
```

Comment:
Should this be enabled by default? (Although I know it is already set as True in our main branch)
I guess most of the users will expect they get the same model if they do not apply schedule to the model, which means the default behavior of `model = hcl.build(sch)` should return exactly the same model as `sch.mod`. This property is not preserved since we reinitialize the weights.

Reply:
I suppose this has no effect if the model already initialized?

Reply:
Not really. Currently if the model has been initialized and `init_weight` is set as `True` by default, it will still reinitialize the parameters.

Reply:
Ok I checked the .build logic and confirmed you're right. Previously Shuai's argument was our current use cases are mainly LLMs so delay initialization is more common, but we could discuss whether there's a better way to cover all cases.